x-pack/test/api_integration/apis/endpoint/alerts.ts

```diff
@@ -165,16 +165,69 @@ export default function({ getService }: FtrProviderContext) {
         expect(body.alerts.length).to.eql(0);
       });
 
-      it('should return 400 when using `before` by custom sort parameter', async () => {
-        await supertest
+      it('alerts api should return data using `before` by custom sort parameter, descending', async () => {
+        const { body } = await supertest
           .get(
-            `/api/endpoint/alerts?${nextPrevPrefixDateRange}&${nextPrevPrefixPageSize}&${nextPrevPrefixOrder}&sort=process.pid&before=1&before=66008e21-2493-4b15-a937-939ea228064a`
+            `/api/endpoint/alerts?${nextPrevPrefixDateRange}&${nextPrevPrefixPageSize}&${nextPrevPrefixOrder}&sort=process.name&before=malware%20writer&before=4d7afd81-26ec-47c0-9741-ae16d331f73d`
           )
           .set('kbn-xsrf', 'xxx')
-          .expect(400);
+          .expect(200);
+        let valid: boolean = true;
+        body.alerts.forEach(alert => {
+          if (alert.process?.name > 'malware writer') {
+            valid = false;
+          }
+        });
+        expect(valid).to.eql(true);
       });
 
-      it('should return data using `after` by custom sort parameter', async () => {
+      it('alerts api should return data using `before` on undefined primary sort values by custom sort parameter, descending', async () => {
+        const { body } = await supertest
+          .get(
+            `/api/endpoint/alerts?${nextPrevPrefixDateRange}&${nextPrevPrefixPageSize}&order=desc&sort=process.parent.entity_id&before=&before=504dc351-5325-45ad-b406-8e4b28c63e2d&empty_string_is_undefined=true`
+          )
+          .set('kbn-xsrf', 'xxx')
+          .expect(200);
+        let lastSeen: string = 'zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz';
+        let valid: boolean = true;
+        body.alerts.forEach(alert => {
+          const entityId = alert.process?.parent?.entity_id;
+          if (entityId === undefined && alert.event.id > '504dc351-5325-45ad-b406-8e4b28c63e2d') {
```

Comment:
I keep seeing this UUID. Is this part of your fixture? Should it be a constant?

Reply:
Why are  you checking if the alert.id is `>` some uuid string?

Reply:
Because it's sorted and I'm asking for all things "after" or "before" this certain thing in the set. Asserting that this is true.

I can make it a constant, sure. It's the only example of a nullable field in our current fixtures, so that's why I re-used it.

Reply:
at some point we'll probably want to regenerate the fixtures and the random data may change.  can we use multiple api calls to avoid tying the test to specific random data?

Reply:
Why would you ever want to sort by a randomly generated uuid?

Reply:
It's not randomly generated at test run time, it's randomly generated when we create the test archive.  When we update the archive in the future, for example upon adding new fields to sample alerts, then we'll have to update these hardcoded values as well.  It would be nice to load the UUIDs from the basic list API instead of hardcoding them.

Reply:
@madirey I agree with @marshallmain. We should try to move away from hard coding alert ids or other alert data in our tests because this data will change every time we decide to update our esArchive. We already went through the pain of updating our tests after updating the esArchive with @dplumlee and it's not fun. I think we can create a tech debt item for us to come up with a solution, but for now could you please create descriptive variables for all the alert ids you use in these tests so it's at least a little easier to update?

Reply:
@alexk307 You wouldn't. But the event ID is a necessary secondary sort to break ties in the primary sort.

Reply:
@madirey Looks good. Thanks.

Reply:
What's the downside to grabbing the UUID's programatically at test time?  i.e. make an api call to `api/endpoint/alerts` and get the first page of alerts, then use the result of that call to test the cursor or whatever else needs a specific UUID in the url?

Reply:
There is no downside. I think that's what we need to do. I would just prefer if we did that in a separate PR.

Reply:
@peluja1012 @marshallmain I mean, in general I agree... but how do you propose I find the one field that has missing values? I can query for it, but I'd have to literally test every field on the alert to find this... these tests are specifically checking sorting and paging on columns that don't always have data...

Reply:
If you have ideas, I'm all ears. I can't think of a way to do it that makes sense.

Reply:
@marshallmain Can we make the assumption that `process.parent.entity_id` will always have at least one empty value?

Reply:
Looking at the fixture data more closely the alerts all have `process.parent.entity_id` so this isn't checking any alerts that have undefined values for the sort field.  We can update the test fixture with some alerts that don't have any ancestor processes so they won't have `process.parent.entity_id` and then make the assumption that the test fixture will have some alerts with and some without that field.

Currently the only documents without a parent are the root process events, but in the test data none of those process events have associated alerts.

Reply:
@marshallmain One of them doesn't have `entity_id` ... it's the one with the event ID indicated...

Reply:
@marshallmain I see... maybe I wasn't looking at an alert?

Reply:
Yeah that event isn't an alert.  The alerts were all created with the same default ancestor settings (3 ancestors) so they all have parents.

Reply:
I think a good approach would be to update the fixture with alerts that don't have ancestors and use the lastSeen value as the tiebreaker in the url.  That should test what we want and break the coupling with the random values.

Reply:
@marshallmain Okie dokes. Did you use the generator to build the fixture? Maybe we could make a change to the generator so that at least one alert has no ancestors?

Reply:
Yeah I used the generator, the easiest way would be to run the generator multiple times with different settings to create different types of tree structures.  Running it with `--ancestors 0` would do it.

Reply:
@marshallmain I manually removed one, but we can do that next time. I think I've addressed your feedback, please take another look. Thanks!